engine/table/src/main/java/io/deephaven/engine/table/impl/locations/impl/AbstractTableLocationProvider.java

```diff
@@ -198,18 +198,18 @@ public TableLocation getTableLocationIfPresent(@NotNull final TableLocationKey t
         // See JavaDoc on tableLocations for background.
         // The intent is to create a TableLocation exactly once to replace the TableLocationKey placeholder that was
         // added in handleTableLocationKey.
-        if (current instanceof TableLocation) {
-            return (TableLocation) current;
-        }
-        synchronized (tableLocations) {
-            current = tableLocations.get(tableLocationKey);
-            if (current instanceof TableLocation) {
-                return (TableLocation) current;
+        while (!(current instanceof TableLocation)) {
+            final TableLocationKey immutableKey = (TableLocationKey) current;
+            // noinspection SynchronizationOnLocalVariableOrMethodParameter
+            synchronized (immutableKey) {
```

Comment:
I'm trying to follow the guarantees this provides. I don't _think_ that `TableLocationKey` are natively guaranteed to be the same exact instance, unless the flow to here guarantees that in some way? I'm assuming we are relying on `tableLocations` to essentially intern the keys for us, yes?

Reply:
TLKs only end up in `tableLocations` via `io.deephaven.engine.table.impl.locations.impl.AbstractTableLocationProvider#handleTableLocationKey`.
That does a `putIfAbsent` of an immutable version of the TLK that we handle.

We only ever swap the TLK with a TL.
So, if we don't find a TL, then the TLK is the unique instance we stored in `handleTLK`.

So, yes, we are effectively interning immutable TLKs, which conveniently makes them good for TLK-level locking.